scripts/release.sh

```diff
@@ -0,0 +1,201 @@
+#!/bin/sh
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# This script will collect some input from user and proceed with bumping versions
+# in crates and libraries, setting up tag and finally bumping versions again.
+#
+# The following is collected from user:
+#  - new version
+#  - the next version
+#
+# The new tag will look like this: "v${newVersion}".
+#
+# Changes to Cargo.toml files will be comitted.
+#
+# Script will not proceed if project is not clean, not on "master" branch or
+# current master differs from remote master.
+#
+# Preprequisites:
+# Before running this script make sure "cargo-release" software in installed:
+# https://github.com/crate-ci/cargo-release
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+line="- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -"
+
+WORK_STATUS=$(git status --porcelain)
+if [ "${WORK_STATUS}" != "" ];
+then
+  echo $line
+  echo "Working directory must be clean. Currently it contains some changes!"
+  echo $WORK_STATUS
+  echo $line
+  exit 1
+fi
+
+# Are we allowed to execute this script on the branch other than master?
+ALLOW_BRANCH="n"
+for param in $1
+do
+  case "$param" in
+    "--allow-branch")
+      ALLOW_BRANCH="yes"
+      echo "Execution of this script on non-master branch is allowed"
+    ;;
+  esac
+done
+CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+if [ "$ALLOW_BRANCH" == "n" ];
+then
+
+  if [ "${CURRENT_BRANCH}" != "master" ];
+  then
+    echo $line
+    echo "Current branch must be master!"
+    echo $CURRENT_BRANCH
+    echo $line
+    exit 1
+  else
+    # refresh master branch
+    echo "Fetching master from remote..."
+    git fetch
+    git pull
+
+    returnStatus=$(($?+0))
+    if [ $returnStatus -ne 0 ];
+    then
+      exit $returnStatus
+    fi
+
+    echo "Comparing current master with remote..."
+    PUSH_STATUS=$(git diff HEAD origin/master --name-status)
+    if [ "${PUSH_STATUS}" != "" ];
+    then
+      echo "There are comitted but not pushed changes in your current branch!!!"
+      exit 1
+    fi
+  fi  # end of when on master
+
+fi # end of allow branch
+
+LATEST_TAG=$(git describe --abbrev=0 --tags)
+LATEST_SEMVER=${LATEST_TAG/v/}
+
+echo $line
+echo ""
+echo "CURRENT_BRANCH=$CURRENT_BRANCH"
+echo "LATEST_TAG=$LATEST_TAG"
+echo "LATEST_SEMVER=$LATEST_SEMVER"
+echo ""
+
+echo "Provide a new semver number. Type: 'a' or just press 'enter' for cancel and abort here."
+echo "Otherwise type version number, for example, type '2.0.0' the tag name will be v2.0.0"
+unset ANSWER
+read ANSWER
+if [ "${ANSWER}" == "a" ] || [ "${ANSWER}" == "" ];
```

Comment:
We create a tag first and then execute publish action. What action do you refer to?

Reply:
- I tried earlier today to run the script, and although I ran into some other error related to napi, this led me to think, everytime I run the script and there are errors, when I re-run it, it's pretty cumbersome to repeat the inputs vs just using arguments

- I think this script is doing too much. This should be run as a pre-publish step with just the below lines
```bash
echo """
Bumping versions of all crates to $NEW_VERSION
  
Bumping versions of the following NPMs:
  - cohort_sdk_client         $NEW_VERSION
  - packages/cohort_sdk_js    $NEW_VERSION
"""

currentDir=$(pwd)
cd cohort_sdk_client
npm version $NEW_VERSION
cd $currentDir
cd packages/cohort_sdk_js
npm version $NEW_VERSION
cd $currentDir
git add --all
git commit -a -m "chore(npm): Release npm $NEW_VERSION" --no-verify

# This will update version in Cargo.toml files and in dependencies, then commit
cargo release --workspace --no-confirm --no-tag --no-publish --no-push -x $NEW_VERSION

returnStatus=$(($?+0))
```

We shouldn't control through the script how the versions are created, we don't need to. Any one should be able to use standard `git tag` command and create a tag and push. 

The `publish` github action should get the versions ready for publishing to `npm` and `cargo` via the above lines of code. 

And finally publish.

Reply:
Sorry you have to explain it in details. The way you described will not work, because `git tag` is not the only step here. 

Lets me say that using of script is optional and all can be done manually. With script it is easier as it makes sure that there are no mistakes in "releasing" from dirty woking tree, or from non-master branch, unless person specifies it explicitly with argument. With script tag annotation and commit annotations become standard, without typos. Version bumps are automated and committed. In the end of the day the release script is bumping versions, committing and tagging. Tag push is manual step of course.

Do you have another release process in mind?

Reply:
No, the script is very useful. But what I am suggesting is we shouldn't enforce creating the tags only using this script. 

So right now, this script is responsible for
> - `NEW VERSION`
>   - Getting the version from user
>   - Bumping versions in `package.json`
>   - Bumping versions in `Cargo.toml`
>   - Create a release commit
>   - Tagging the repo
> -`NEXT VERSION`
>   - Getting the version from user
>   - Bumping versions in `package.json`
>   - Bumping versions in `Cargo.toml`
>   - Create a release commit


What I was suggesting was:
> - `NEW VERSION`
>   - Getting the version from user
>   - Bumping versions in `package.json`
>   - Bumping versions in `Cargo.toml`
>   - Create a release commit
>   - Tagging the repo
>   - Push the commit and tag to remote.

This way, we can use the script in as part of github actions in `publish.yaml` as the first step before the publishing step.